src/lib.rs

```diff
@@ -185,8 +169,39 @@ pub fn parse_path(path: &PathBuf, config: ParseConfig) -> anyhow::Result<ParseOu
         stats.ok += 1;
 
         // lol this clone, probably shouldn't use entry
+        // TODO: output should be able to generate this without explicitly creating
         let compile_directory = directory.entry(e.compile_id.clone()).or_default();
 
+        for parser in &all_parsers {
+            if let Some(md) = parser.get_metadata(&e) {
+                let results = parser.parse(lineno, md, e.rank, &e.compile_id, &payload);
+                match results {
+                    Ok(results) => {
+                        for (filename, out) in results {
+                            output.push((filename.clone(), out));
+                            compile_directory.push(filename);
+                        }
+                    }
+                    Err(err) => {
+                        match  parser.name() {
+                            "dynamo_guards" => {
+                                eprintln!("Failed to parse guards json: {}", err);
+                                stats.fail_dynamo_guards_json += 1;
+                            }
+                            name => {
+                                eprintln!("Parser {name} failed: {err}");
+                                stats.fail_parser += 1;
+                            }
+                        }
+                    }
+                }
+
+            }
+        }
+
+        // TODO: implement this as StructuredLogParser
```

Comment:
I think the abstraction you've introduced is good for artifacts, but I think there's a class of other things that we could potentially get from the trace log which don't create a file. For example, imagine if we started emitting traditional traces (start + end events) to the log, and then wanted to visualize these together. These would be strewn across many different trace messages and you wouldn't want to make a file per each one.

I think my primary ask here is to avoid overabstracting, at least for now. I think there's a decent case to be made for generalizing artifacts (but note that artifacts as currently implemented have some funny problems, e.g., when ddp optimize is on, you'll get multiple copies of the same artifact in the same compile id), but I would hesitate to say that we have an abstraction that works for arbitrary analyses you might want to do. (If you *really* wanted to design it, you'd probably want some sort of state machine per analysis, with enough structure in the input parsing so you can efficiently dispatch to the correct analyses that actually care about a given token without having to O(N) loop through all analyses... and that's not getting into if there's ever shared information that wants to be saved over other analyses. It probably also matters if we want streaming or if we can just assume everything fits in RAM.)

Reply:
Yeah, it might be that analyses that need to span across multiple envelopes and events, such as this one, won't fit into this model, which is fine. I think the case I want to cover is the most common one where someone wants to log a some type of event that occurs a constant number of times per compilation, that can be rendered per event (i.e. compilation metrics)

Trying to abstract the idea of "log collects global information into a single template or UI" like the one for the stack trie gets to the point of being cumbersome